amaranth/lib/io.py

```diff
@@ -253,6 +258,362 @@ def __repr__(self):
+        o: :py:`unsigned(width)` (if :py:`direction in (Direction.Output, Direction.Bidir)`)
+        oe: :py:`unsigned(1)` (if :py:`direction in (Direction.Output, Direction.Bidir)`)
+        """
+        def __init__(self, direction, width):
+            self._direction = Direction(direction)
+            self._width = operator.index(width)
+            members = {}
+            if self._direction is not Direction.Output:
+                members["i"] = wiring.Out(self._width)
+            if self._direction is not Direction.Input:
+                members["o"] = wiring.In(self._width)
+                members["oe"] = wiring.In(1, init=int(self._direction is Direction.Output))
+            super().__init__(members)
+
+        @property
+        def direction(self):
+            return self._direction
+
+        @property
+        def width(self):
+            return self._width
+
+        def __eq__(self, other):
+            return type(self) == type(other) and self.direction == other.direction and self.width == other.width
+
+        def __repr__(self):
+            return f"Buffer.Signature({self.direction}, {self.width})"
+
+    def __init__(self, direction, port):
+        if not (hasattr(port, "direction") and isinstance(port.direction, Direction)):
+            raise TypeError(f"'port' must be a 'SingleEndedPort', 'DifferentialPort', "
+                            f"or a platform-specific object, not {port!r}")
+        self._port = port
+        super().__init__(Buffer.Signature(direction, len(port)))
+        if port.direction is Direction.Input and self.direction is not Direction.Input:
+            raise ValueError(f"Input port cannot be used with {self.direction.name} buffer")
+        if port.direction is Direction.Output and self.direction is not Direction.Output:
+            raise ValueError(f"Output port cannot be used with {self.direction.name} buffer")
+
+    @property
+    def port(self):
+        return self._port
+
+    @property
+    def direction(self):
+        return self.signature.direction
+
+    def elaborate(self, platform):
+        if hasattr(platform, "get_io_buffer"):
+            res = platform.get_io_buffer(self)
+            if res is not None:
+                return res
+
+        m = Module()
+
+        invert = sum(bit << idx for idx, bit in enumerate(self._port.invert))
+        if self.direction is not Direction.Input:
+            if invert:
+                o_inv = Signal.like(self.o)
+                m.d.comb += o_inv.eq(self.o ^ invert)
+            else:
+                o_inv = self.o
+        if self.direction is not Direction.Output:
+            if invert:
+                i_inv = Signal.like(self.i)
+                m.d.comb += self.i.eq(i_inv ^ invert)
+            else:
+                i_inv = self.i
+
+        if isinstance(self._port, SingleEndedPort):
+            if self.direction is Direction.Input:
+                m.submodules.iob = IOBufferInstance(self._port.io, i=i_inv)
```

Comment:
`io_buffer` or `buffer` or `inner` or something maybe? "iob" is confusing to newcomers, and just about everyone will see this in the hierarchy.

Reply:
this doesn't actually show up in the hierarchy (it's lowered to an anonymous `$tribuf`)

Reply:
Oh. Then I guess it can be just `m.submodules +=`?